src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
You are adding a new piece of code. I think it, at least, should be tested for correct behavior under conditions that I described. Let's add tests for scenarios that go through this code path with types that differ only in dynamic or tuple name mismatch (both not limited to the top level) and confirm that result is what we would expect to get.

---
In reply to: [100627591](https://github.com/dotnet/roslyn/pull/16928#discussion_r100627591) [](ancestors = 100627591,100616402,100602791,100601157,100591895)